### DISCUSSION TITLE
How to create PersistantVolume (ReadWriteOnce) on the node that runs the workflow? (Kubernetes backend)

### DISCUSSION THREAD

**Discussion by itytophile** [woodpecker-ci/woodpecker#3068] [Ask a Question] [answered]
Hi. I am running Woodpecker on a little k3s cluster. The workflows run correctly, however they always run on the same node (I have two nodes).

I've set the `nodeSelector` option in `woodpecker.yml` to force the workflow to run on the node I want. The newly created pod wants to run on the good node but the PersistantVolume is created on the other node. The PersistentVolume uses the ReadWriteOnce access mode, so the workflow pod can't access it and stays "pending" (they are not on the same node).

Pending pod description:

```
Node-Selectors:              kubernetes.io/arch=arm64
                             merde=merde
Events:
  Type     Reason            Age                  From               Message
  ----     ------            ----                 ----               -------
  Warning  FailedScheduling  19m                  default-scheduler  0/2 nodes are available: 1 node(s) didn't match Pod's node affinity/selector. preemption: 0/2 nodes are available: 1 No preemption victims found for incoming pod, 1 Preemption is not helpful for scheduling..
  Warning  FailedScheduling  9m35s (x2 over 14m)  default-scheduler  0/2 nodes are available: 1 node(s) didn't match Pod's node affinity/selector. preemption: 0/2 nodes are available: 1 No preemption victims found for incoming pod, 1 Preemption is not helpful for scheduling..
```

From the PersistantVolume:
```
Node Affinity:     
  Required Terms:  
    Term 0:        kubernetes.io/hostname in [not-the-good-node]
```

Good node description:
```
Labels:             beta.kubernetes.io/arch=arm64
                    beta.kubernetes.io/instance-type=k3s
                    beta.kubernetes.io/os=linux
                    has-space=true
                    kubernetes.io/arch=arm64
                    kubernetes.io/hostname=the-good-node
                    kubernetes.io/os=linux
                    merde=merde
                    node.kubernetes.io/instance-type=k3s
```

My `woodpecker.yml`:
```yaml
steps:
  coucou:
    image: busybox:musl
    commands:
      - echo coucou
    backend_options:
      kubernetes:
        resources:
          requests:
            memory: 100Mi
            cpu: 100m
        nodeSelector:
          merde: merde
```

**Comment by zc-devs** [comment]
1. Do you use Rancher's Local Path Provisioner?
2. How many Woodpecker Agents do you use?
2.1 If one, which node does it run on?
4. How many [Workflows per agent](https://woodpecker-ci.org/docs/administration/agent-config#workflows-per-agent) is set?
5. Make sure that [Cancel previous pipelines](https://woodpecker-ci.org/docs/usage/project-settings#cancel-previous-pipelines) is not set.
6. Make sure you are running a few pipelines simultaneously.

**Comment by itytophile** [comment]
1. Yes I do
2. Two, they both run on the same node (the good node)
3. One
4. Done
5. Done

**Comment by zc-devs** [comment]
2. It is strange. They should spawn across nodes. Can you test the case when Agents run on different nodes? Can you attach also full definition of Pods, PV, PVC as files? Especially I want to look at `volumeBindingMode`, but other info might helpful also.

**Comment by itytophile** [comment]
That's on me. There is a nodeSelector (I was testing things) on the agent pods that's why they are on the same node. I'll change that and retry running a workflow.

**Comment by itytophile** [comment]
Here are the files. With one agent on each node. I didn't edit the nodes name this time: the "good" node is "renegade-one" and the other one is "nanopi.lan". The secret has been edited.
[workflow-pod.txt](https://github.com/woodpecker-ci/woodpecker/files/13810783/workflow-pod.txt)
[server.txt](https://github.com/woodpecker-ci/woodpecker/files/13810784/server.txt)
[pvc.txt](https://github.com/woodpecker-ci/woodpecker/files/13810785/pvc.txt)
[pv.txt](https://github.com/woodpecker-ci/woodpecker/files/13810786/pv.txt)
[agent-2.txt](https://github.com/woodpecker-ci/woodpecker/files/13810787/agent-2.txt)
[agent-1.txt](https://github.com/woodpecker-ci/woodpecker/files/13810788/agent-1.txt)

**Comment by zc-devs** [comment, marked as answer]
K3s `v1.29.0+k3s1`, 3 nodes
Woodpecker `2.2.2`, 1 server, 1 agent with max workflows 4

I ran pipeline four times
![Screenshot 2024-01-26 1](https://github.com/woodpecker-ci/woodpecker/assets/127358482/6d1d9996-4d9e-4075-8402-4bd9f565a5cb)
and workflows were spawned across three nodes
![Screenshot 2024-01-26 2](https://github.com/woodpecker-ci/woodpecker/assets/127358482/6df75792-71ff-4630-8eee-d10442312745)
At start of pipeline workspace volume is created. Then repository cloned into it. Then it mounts into each step also.
Volume is created on node, where first pod runs (binding). Rest of the pods are scheduled on the same node due to affinity, that LocalPathProvisioner sets.
So, seems works as intended.

> I've set the nodeSelector option in woodpecker.yml to force the workflow to run on the node I want. The newly created pod wants to run on the good node but the PersistantVolume is created on the other node.

That also seems logical. Local volume was created on one node. You created pod on another. Pod cannot mount _local_ volume from first node.

> they always run on the same node (I have two nodes)

And you have two agents. Looks like an issue with Server->Agent scheduling.

> - Is there some documentation on how tasks are scheduled to agents? What's the algorithm? Can I assign some priorities on agents other than select by labels or exclude agent from taking tasks completely?
> - Well priorizing is an open issue 😅 The algo is fifo

Maybe related: [Kubernetes: smart(er) agent affinity](https://github.com/woodpecker-ci/woodpecker/issues/3070)

You supposed to be able to disable Agents, but [agent option "stop agent from taking new tasks" not working](https://github.com/woodpecker-ci/woodpecker/issues/1990). I just delete unnecessary Agent, run pipeline via the second and then restart the first.

In your case, it is, probably, simpler to use one agent and give away scheduling to Kubernetes.
